### PR TITLE
Refactor `app/gui/queryresults/queryresultsservice.js`

### LINKED ISSUE #58: Refactor and rename `QueryResultsService.addRemoveFeaturesToLayerResult()` function
Usually such ambiguous (add + remove = ? ) and long names denote that a function that is too large (and quite difficult to debug):

https://github.com/g3w-suite/g3w-client/blob/e8996a7c6e9bacc00c4b5c2504d87b92e7d7b5f0/src/app/gui/queryresults/queryresultsservice.js#L298-L345

Please whenever possible consider breaking the function down into simpler subroutines (and using shorter names).

### LINKED ISSUE #59: Slim down `QueryResultsService.setActionsForLayers()` function
The following function is nearly 250 lines long:

https://github.com/g3w-suite/g3w-client/blob/e8996a7c6e9bacc00c4b5c2504d87b92e7d7b5f0/src/app/gui/queryresults/queryresultsservice.js#L386-L634

If possible, consider simplifying it by using subroutines.

### PR DESCRIPTION
**Pull request by Raruto**
## List of changes:

- add tons `@FIXME` comments (eg. please `"add function description"` here)
- code cleanup (eg. tons of weird [ternary conditions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator), [short-circuiting](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_AND#short-circuit_evaluation) logic and missing [brackets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/if...else)!)
- full function rewrite `QueryResultsService::updateLayerResultFeatures(layer)`
- notable function rewrite `QueryResultsService::downloadFeatures(type, layer, features=[], action, index)`
- notable function rewrite `QueryResultsService::_addRemoveSelectionFeature(layer, feature, index, force)`
- notable function rewrite `QueryResultsService::setActionsForLayers(layers, options = { add: false })`
- removed function `QueryResultsService::_digestFeaturesForLayers(featuresForLayers)` (replaced by `QueryResultsService::setQueryResponse`)
- deprecate function `QueryResultsService.prototype.downloadApplicationWrapper(downloadFnc, options)` in favor of `GUI::downloadWrapper(downloadFnc, options)`

Closes: https://github.com/g3w-suite/g3w-client/issues/58
Closes: https://github.com/g3w-suite/g3w-client/issues/59

Preparatory for: https://github.com/g3w-suite/g3w-client/issues/84
Related to: https://github.com/g3w-suite/g3w-client/issues/193

---

~~**NB** this pull depends on https://github.com/g3w-suite/g3w-client/pull/308 for the addition of the [`"util-deprecate"`](https://www.npmjs.com/package/util-deprecate) dependency (so it is recommended to merge the other one first to keep things tidy)~~

### CODE REVIEW

File: src/app/gui/queryresults/queryresultsservice.js [hunk 1]
```diff
@@ -216,245 +396,233 @@ inherit(QueryResultsService, G3WObject);
...
+  if (1 === this.state.layers.length) {
+    this.highlightFeaturesPermanently(this.state.layers[0]);
+  }
 
-/**
- * @deprecated since v3.8. Will be deleted in 4.x. Use QueryResultsService::updateLayerResultFeatures(layer) instead
- */
-proto.addRemoveFeaturesToLayerResult = proto.updateLayerResultFeatures;
+  this.changeLayerResult(layer);
+};
```
Comment (Raruto) [inline src/app/gui/queryresults/queryresultsservice.js]: this function has changed substantially, check it carefully

File: src/app/gui/queryresults/queryresultsservice.js [hunk 2]
```diff
@@ -889,93 +1142,147 @@ proto.clearState = function(options={}) {
...
 /**
- * Method that convert response from Data Provider to a Query Result component data structure
+ * Converts response from DataProvider into a QueryResult component data structure
+ * 
  * @param featuresForLayers: Array contains for each layer features
+ * 
  * @returns {[]}
- * @private
  */
 proto._digestFeaturesForLayers = function(featuresForLayers) {
```
Comment (Raruto) [inline src/app/gui/queryresults/queryresultsservice.js]: this is a blob function, I partially rewrote it to try to make it more readable, check it carefully

File: src/app/gui/queryresults/queryresultsservice.js [hunk 3]
```diff
@@ -1341,99 +1697,112 @@ proto.printAtlas = function(layer, feature){
...
+    if (features.length <= 1) {
+      this.setCurrentActionLayerFeatureTool({ index, action, layer });
     }
   };
-  if (query.type === 'polygon'){
-    //check if multidownload if present
+
+  if ('polygon' !== query.type) {
+    runDownload();
+  } else { // check if multi-download if present
```
Comment (Raruto) [inline src/app/gui/queryresults/queryresultsservice.js]: within this function `QueryResultsService::downloadFeatures(type, layer, features=[], action, index)`, i rewrote several if-else conditions to reduce nesting level, check carefully

File: src/app/gui/queryresults/queryresultsservice.js [hunk 4]
```diff
@@ -1447,342 +1816,428 @@ proto.downloadFeatures = async function(type, layer, features=[], action, index)
...
+      this.setLayerActionTool({ layer });
+    }
+
+    /** @FIXME add description */
+    if (features.length < 1 && undefined !== downloadsactions) {
+      this.setLayerActionTool({ layer, component: QueryPolygonCsvAttributesComponent, config });
     }
-  } else runDownload();
+
+  }
```
Comment (Raruto) [inline src/app/gui/queryresults/queryresultsservice.js]: within this function `QueryResultsService::downloadFeatures(type, layer, features=[], action, index)`, i rewrote several if-else conditions to reduce nesting level, check carefully

File: src/app/gui/queryresults/queryresultsservice.js [hunk 5]
```diff
@@ -1447,342 +1816,428 @@ proto.downloadFeatures = async function(type, layer, features=[], action, index)
...
+
+      /** @FIXME add description */
+      if (1 === this.state.layers.length && !this.state.layers[0].features.length) {
+        this.state.layers.splice(0);
+      }
+
     }
+
   }
+
```
Comment (Raruto) [inline src/app/gui/queryresults/queryresultsservice.js]: within this function `QueryResultsService::_addRemoveSelectionFeature(layer, feature, index, force)`, i rewrote several if-else conditions to reduce nesting level, check carefully

File: src/app/gui/queryresults/queryresultsservice.js [hunk 6]
```diff
@@ -1,69 +1,205 @@
...
+  /**
+   * @FIXME add description
+   */
   this.unlistenerlayeractionevents = [];
+
+  /**
+   * @FIXME add description
+   */
   this._actions = {
     'zoomto': QueryResultsService.zoomToElement,
```
Comment (Raruto) [inline src/app/gui/queryresults/queryresultsservice.js]:
```js
QueryResultsService.zoomToElement
```

is this an undefined function?

Comment (volterra79) [reply src/app/gui/queryresults/queryresultsservice.js]: @Raruto yes. You can delete it. Thanks

Comment (Raruto) [reply src/app/gui/queryresults/queryresultsservice.js]: removed in https://github.com/g3w-suite/g3w-client/pull/317/commits/40d0254fb7a5d2bf20e59990c20cde13f92694df

File: src/app/gui/queryresults/queryresultsservice.js [hunk 7]
```diff
@@ -1,1874 +1,2555 @@
...
+
+      try {
+        await CatalogLayersStoresRegistry.getLayerById(layer.id).getDownloadFilefromDownloadDataType(type, { data }) || Promise.resolve();
+      } catch(err) {
+        GUI.notify.error(err || t("info.server_error"));
+      }
+
+      ApplicationService.setDownload(false, download_caller_id);
+
+      GUI.setLoadingContent(false);
```
Comment (Raruto) [inline src/app/gui/queryresults/queryresultsservice.js]: @volterra79 after, try to see if you are able to rewrite these kind of statements to make use of some sort of download wrapper (eg. using some other `utils` function I'm not aware of):

```suggestion
      await this.downloadApplicationWrapper(
        CatalogLayersStoresRegistry
          .getLayerById(layer.id)
          .getDownloadFilefromDownloadDataType
          .bind(this, type, { data })
      );
```

**Review by volterra79 — APPROVED** [review]

**Pull request merged**